File: src/main/java/com/company/service/CourseService.java
package com.company.service;

import com.company.entities.Course;
import com.company.repository.CourseRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.stream.Collectors;

@Service
public class CourseService {

    @Autowired
    private CourseRepository repository;
    @Autowired
    private StudyGroupService groupService;

    public List<Course> findAll(){
        List<Course> courses = (ArrayList<Course>)repository.findAll();
        return courses.stream().sorted(Comparator.comparing(Course::getId)).collect(Collectors.toList());
    }

    public void addCourse(Course course){
        repository.save(course);
    }

    public Course findCourseById(long id){
        Course course = repository.findById(id).get();
        return course;
    }
    @Transactional
    public void removeCourse(long id){
        findCourseById(id).getGroups().forEach(group -> groupService.removeStudyGroup(group.getId()));
        repository.deleteById(id);
    }

    public void updateCourse(Course course){
        repository.save(course);
    }

}

File: src/main/resources/static/app/service/course.service.js
angular
    .module("SchoolApp")
    .factory("CourseService", ["$http", function ($http) {
        var ALL_COURSES = "http://localhost:8080/course/findAll";
        var ADD_COURSE = "http://localhost:8080/course/add";
        var UPDATE_COURSE = "http://localhost:8080/course/update";
        var DELETE_COURSE = "http://localhost:8080/course/remove/";

        var service = {};


        service.getAllCourses = function getAllCourses() {
            return $http.get(ALL_COURSES);
        };

        service.addCourse = function addCourse(course) {
            return $http.post(ADD_COURSE, course)
        };

        service.updateCourse = function updateCourse(course) {
            return $http.post(UPDATE_COURSE, course)
        };

        service.deleteCourse = function deleteCourse(course) {
            return $http.delete(DELETE_COURSE + course.id)
        };

        return service;
    }]);